clang/lib/3C/AVarBoundsInfo.cpp

```diff
@@ -61,15 +61,434 @@ void AVarBoundsStats::print(llvm::raw_ostream &O,
+            if (CKVar != nullptr && CKVar->isNumConstant())
+              TmpBKeys.insert(CK);
+          }
+          InferredKBnds[INB.first] = TmpBKeys;
+        }
+      }
+    }
+
+    // Now from the newly inferred bounds i.e., InferredKBnds, check
+    // if is is different from previously known bounds of K
+    for (auto &IKB : InferredKBnds) {
+      bool Handled = false;
+      if (CurrIterInferBounds.find(K) != CurrIterInferBounds.end()) {
+        auto &BM = CurrIterInferBounds[K];
+        if (BM.find(IKB.first) != BM.end()) {
+          Handled = true;
+          if (BM[IKB.first] != IKB.second) {
+            BM[IKB.first] = IKB.second;
+            if (IKB.second.empty())
+              BM.erase(IKB.first);
+            IsChanged = true;
+          }
+        }
+      }
+      if (!Handled) {
+        CurrIterInferBounds[K][IKB.first] = IKB.second;
+        if (IKB.second.empty()) {
+          CurrIterInferBounds[K].erase(IKB.first);
+        } else {
+          IsChanged = true;
+        }
+      }
+    }
+  } else if (ErrorOccurred) {
+    // If any error occurred during inferring bounds then
+    // remove any previously inferred bounds for K.
+    IsChanged = CurrIterInferBounds.erase(K) != 0;
+  }
+  return IsChanged;
+}
+bool AvarBoundsInference::inferBounds(BoundsKey K, AVarGraph &BKGraph, bool FromPB) {
+  bool IsChanged = false;
+
+  if (BI->InvalidBounds.find(K) == BI->InvalidBounds.end()) {
+    // Infer from potential bounds?
+    if (FromPB) {
+      auto &PotBDs = BI->PotentialCntBounds;
+      if (PotBDs.find(K) != PotBDs.end()) {
+        ProgramVar *Kvar = BI->getProgramVar(K);
+        std::set<BoundsKey> PotentialB;
+        PotentialB.clear();
+        for (auto TK : PotBDs[K])
+          getReachableBoundKeys(Kvar->getScope(), TK, PotentialB, BKGraph, true);
+
+        if (!PotentialB.empty()) {
+          bool Handled = false;
+          // Potential bounds are always count bounds.
+          // We use potential bounds
+          ABounds::BoundsKind PotKind = ABounds::CountBoundKind;
+          if (CurrIterInferBounds.find(K) != CurrIterInferBounds.end()) {
+            auto &BM = CurrIterInferBounds[K];
+            // If we have any inferred bounds for K then ignore potential
+            // bounds.
+            for (auto &PosB : BM) {
+              if (!PosB.second.empty()) {
+                Handled = true;
+                break;
+              }
+            }
+          }
+          if (!Handled) {
+            CurrIterInferBounds[K][PotKind] = PotentialB;
+            IsChanged = true;
+          }
+        }
+      }
+    } else {
+      // Infer from the flow-graph.
+      std::set<BoundsKey> TmpBkeys;
+      // Try to predict bounds from predecessors.
+      BKGraph.getPredecessors(K, TmpBkeys);
+      IsChanged = predictBounds(K, TmpBkeys, BKGraph);
+    }
+  }
+  return IsChanged;
+}
+
 bool AVarBoundsInfo::isValidBoundVariable(clang::Decl *D) {
```

Comment:
Suggested change for this function.  Why is `VarDecl` different?
```c++
bool AVarBoundsInfo::isValidBoundVariable(clang::Decl *D) {
  // All parameters, return, and field values are valid bound variables.
  if (isa<ParmVarDecl>(D) || isa<FunctionDecl>(D) || isa<FieldDecl>(D))
    return true;
 
 // Why are VarDecls different? 
  if (auto *VD = dyn_cast<VarDecl>(D))
    return !VD->getNameAsString().empty();

  return false;
}
```

Reply:
Want to filter out any dummy VarDecls.